lib/kinesis_client.rb

```diff
@@ -84,19 +78,31 @@ def push_batch(batch)
 
     return_message = {
       failures: resp.failed_record_count,
-      error_messages: resp.records.map {|record| record.error_message }.compact
+      failures_data: filter_failures(resp),
+      # error_messages: resp.records.map { |record| record.error_message }.compact
     }
 
     $logger.info("Message sent to #{config[:stream_name]} #{return_message}") if $logger
 
-    return {
+    {
       "code": "200",
       "message": return_message.to_json
     }
   end
 
   def push_records
-    @batch.each_slice(@batch_size) {|slice| push_batch slice}
-    @batch = []
+    @records.each_slice(@batch_size) { |slice| push_batch(slice) } if @records.length > 0 
+    @records = []
+  end
+
+  def filter_failures(resp)
+    failed_records = []
+    resp.records.each_with_index do |record, i|
+      if record.key?(:error_message)
+        decoded_record = avro.decode(@records[i])
```

Comment:
Not sure this is going to match the records correctly. Let me know if I have this mixed up, but I think `resp` is just one batch, whereas `@records` would be all of them. So if there were more than one batch, e.g. you could have records 0-9, and two batches with records 0-4 and 5-9, for the second batch the indices in `resp` would be `0-4`. So you might have to pass the batch as well as the response to this method.

There should probably be a test case for when we push more than the batch size (see below).